include/seqan3/search/kmer_index/detail/shape_iterator.hpp

```diff
@@ -0,0 +1,461 @@
+        //!\cond
+        requires std::RandomAccessIterator<it_t>
+        //!\endcond
+    {
+        hash_backward(skip);
+        return *this;
+    }
+
+    /*!\anchor shape_iterator_operator-decrement
+     * \brief Return decremented copy of this iterator.
+     * \attention This function is only avaible if it_t models std::RandomAccessIterator.
+     */
+    shape_iterator operator-(difference_type const skip) const noexcept
+        //!\cond
+        requires std::RandomAccessIterator<it_t>
+        //!\endcond
+    {
+        shape_iterator tmp{*this};
+        return tmp -= skip;
+    }
+
+    /*!\brief Non-member operator- delegates to non-friend operator-.
+     * \attention This function is only avaible if it_t models std::RandomAccessIterator.
+     */
+    friend shape_iterator operator-(difference_type const skip, shape_iterator const & it) noexcept
+        //!\cond
+        requires std::RandomAccessIterator<it_t>
+        //!\endcond
+    {
+        return it - skip;
+    }
+
+    /*!\anchor shape_iterator_operator-difference
+     * \brief Return offset between this and remote iterator's position.
+     * \attention This function is only avaible if it_t models std::RandomAccessIterator.
+     */
+    difference_type operator-(shape_iterator const lhs) const noexcept
+        //!\cond
+        requires std::RandomAccessIterator<it_t>
+        //!\endcond
+    {
+        return static_cast<difference_type>(text_right - lhs.text_right);
+    }
+
+    /*!\brief Move the iterator to a given position and return the corresponding hash value.
+     * \attention This function is only avaible if it_t models std::RandomAccessIterator.
+     */
+    value_type operator[](std::make_unsigned_t<difference_type> const n)
+        //!\cond
+        requires std::RandomAccessIterator<it_t>
+        //!\endcond
+    {
+        text_left = std::ranges::next(text_start, n);
+        hash_full();
+        return operator*();
+    }
+
+    //!\brief Return the hash value.
+    value_type operator*() const noexcept
+    {
+        return hash_value + to_rank(*text_right);
+    }
+    
+private:
+    //!\brief The alphabet type of the passed iterator.
+    using alphabet_t = value_type_t<it_t>;
+
+    //!\brief The alphabet size.
+    static constexpr size_t const sigma{alphabet_size_v<alphabet_t>};
+
+    //!\brief The hash value.
+    size_t hash_value{0};
+
+    //!\brief The seqan3::shape to use for hashing.
+    shape shape_;
+
+    //!\brief The factor for the left most position of the hash value.
+    size_t roll_factor{0};
+
+    //!\brief True if `shape_` contains only 1's at [0, k-1] (ungapped), false otherwise.
+    bool ungapped_shape{false};
+
+    //!\brief Iterator pointing to begin of the underlying text.
+    it_t text_start;
+
+    //!\brief Iterator to the leftmost position of the k-mer.
+    it_t text_left;
+
+    //!\brief Iterator to the rightmost position of the k-mer.
+    it_t text_right;
```

Comment:
This iterator is quite big :open_mouth: 

1. Please evaluate if you can reduce the size_t variables, e.g. what is the possibly largest value of `roll_factor`?
1. Can't you encode `ungapped_shape` within the shape?
1. Using the small_vector as base class is convenient, but inefficient, it is 33 bytes, although it could be done <=5 bytes. We need a constexpr bitset :expressionless: 
1. Do you really need text_start? Keep in mind that every iterator is 8bytes
1. We don't need the shape inside the iterator and inside the view. Ideally we would have a bitset, and than it would be more efficient to have it in the iterator, but if we don't have that, maybe we should have it jus tin the the view for now and use a pointer to the view.

Reply:
> This iterator is quite big 😮
> 
> 1. Please evaluate if you can reduce the size_t variables, e.g. what is the possibly largest value of `roll_factor`?
dna4 32-mer = `2^64`, so `size_t` seems fine
> 2. Can't you encode `ungapped_shape` within the shape?
Do you mean within the small_vector or within the class?
> 3. Using the small_vector as base class is convenient, but inefficient, it is 33 bytes, although it could be done <=5 bytes. We need a constexpr bitset 😑
I had an array before, but since we want to replace it anyway, it doesn't matter too much?
> 4. Do you really need text_start? Keep in mind that every iterator is 8bytes
If we want to model random access ... what would `operator[]` do if you only have your current position?
> 5. We don't need the shape inside the iterator and inside the view. Ideally we would have a bitset, and than it would be more efficient to have it in the iterator, but if we don't have that, maybe we should have it jus tin the the view for now and use a pointer to the view.
Mhm, I see the problem, but if we want to change it anyways once the bitset is there, we may as well leave it for now....

Reply:
> > Do you really need text_start? Keep in mind that every iterator is 8bytes
> If we want to model random access ... what would operator[] do if you only have your current position?

Hm I really wonder if random access makes sense for hashing a sequence since the usual use case is to hash from start to end..

Reply:
We should optimise the use-case for ForwardRange access, but it doesn't hurt to have RandomAccess.